### PR TITLE
Backlog 7929

### PR DESCRIPTION
**Pull request by EgorZhuk**
Hello @tmorgner, I have a couple of thingth for review:

1. Clean futures and tasks in a proper way
2. Move auto-scheduling to server side
3. Implement simple UI to view large paginated reports without hanging a browser (see video)

https://drive.google.com/file/d/0BzVckAbDytTXNU5FOXNqODJFVWM/view

### CODE REVIEW

File: test-src/org/pentaho/reporting/platform/plugin/async/PentahoAsyncExecutionTest.java [hunk 1]
```diff
@@ -117,7 +120,7 @@ public void testListenerFailExecution() throws Exception {
 
     PentahoAsyncReportExecution exec = createMockCallable();
     // simulates queuing
-    exec.notifyTaskQueued( UUID.randomUUID() );
+    exec.notifyTaskQueued( UUID.randomUUID(), null );
```
Comment (tmorgner) [inline test-src/org/pentaho/reporting/platform/plugin/async/PentahoAsyncExecutionTest.java]: This call should actually error out. If you have a list parameter, never pass in Null, demand that your API user passes in Collections.emptyList() instead.

File: src/org/pentaho/reporting/platform/plugin/async/AsyncReportStatusListener.java [hunk 2]
```diff
@@ -52,11 +56,13 @@
...
   public AsyncReportStatusListener( final String path,
-                             final UUID uuid,
-                             final String mimeType ) {
+                                    final UUID uuid,
+                                    final String mimeType,
+                                    final List<? extends ReportProgressListener> callbackListeners ) {
     this.path = path;
     this.uuid = uuid;
     this.mimeType = mimeType;
+    this.callbackListeners = callbackListeners;
```
Comment (tmorgner) [inline src/org/pentaho/reporting/platform/plugin/async/AsyncReportStatusListener.java]: Always check your list parameters for null. Dont allow it, demand that they pass in an empty list instead. Then you dont have to check for null all the time.

Do you expect that users are allowed to modify that list after it has been set? If not, make a copy for your private use to isolate yourself from evil user actions.

**Comment by EgorZhuk** [comment]
Another thing we need here - replace CSS links with content to make it work when user downloads the file.

**Comment by tmorgner** [comment]
There are conflicts. Can you rebase this agains the latest EWOK branch?

**Comment by EgorZhuk** [comment]
@tmorgner done. Also bukk check added. I'm working on embedded css and images now.

File: test-src/org/pentaho/reporting/platform/plugin/async/AsyncReportStatusListenerTest.java [hunk 3]
```diff
@@ -1,7 +1,9 @@
 package org.pentaho.reporting.platform.plugin.async;
```
Comment (tmorgner) [inline test-src/org/pentaho/reporting/platform/plugin/async/AsyncReportStatusListenerTest.java]: No license header.

File: src/org/pentaho/reporting/platform/plugin/output/PaginationControlWrapper.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
+package org.pentaho.reporting.platform.plugin.output;
```
Comment (tmorgner) [inline src/org/pentaho/reporting/platform/plugin/output/PaginationControlWrapper.java]: No license header.

**Comment by tmorgner** [comment]
Only some license headers missing, then I can merge.

**Comment by EgorZhuk** [comment]
@tmorgner added licenses, also added code to embed css. It works fine with styles but the same trick with images it too slow :(

**Comment by EgorZhuk** [comment]
@tmorgner I just checked how platform scheduler works - it places css and image in export folder and replaces links wih "api/repo/foo". I believe it's not what we want because we will have css for every page and numerous images.

**Comment by EgorZhuk** [comment]
@tmorgner

**Comment by EgorZhuk** [comment]
@tmorgner I see two ways of making images work:

1) Do it like in a platform scheduler - it will require basic auth if user wants to see pictures
2) Do it well - add images to the archive that user downloads to his machine

**Pull request merged**